torch_xla/csrc/tensor.cpp

```diff
@@ -630,29 +630,24 @@ std::string XLATensor::DumpHloComputation(
                             : std::string();
 }
 
-std::shared_ptr<XLATensor::ShardingSpec> XLATensor::sharding_spec() const {
-  XLA_CHECK(data()->sharding_spec != nullptr)
-      << "Trying to access a null cursor";
-  return data()->sharding_spec;
-}
-bool XLATensor::IsShardingAnnotated() const {
-  return data()->sharding_spec != nullptr;
-}
-void XLATensor::SetShardingSpec(const xla::OpSharding& sharding,
-                                bool replicated, bool manual) {
-  auto new_sharding_spec =
-      std::make_shared<ShardingSpec>(sharding, replicated, manual);
-  data()->sharding_spec = new_sharding_spec;
-  XLA_CHECK(data()->ir_value.node != nullptr)
-      << "Tyring to access a null cursor";
+XLATensor::ShardingSpecPtr XLATensor::sharding_spec() const {
+  XLA_CHECK(GetIrValue().node != nullptr) << "Tyring to access a null cursor";
+  auto* sharding =
+      dynamic_cast<XlaNode*>(data()->ir_value.node.get())->GetSharding();
+  if (sharding == nullptr) {
+    return nullptr;
+  }
+  return std::make_shared<ShardingSpec>(*sharding);
```

Comment:
why do you need shardingspec to be a shared_ptr when you store it as raw_ptr in ir?

Reply:
We store it as a shared ptr in IR, and I think we should prefer returning a shared ptr instead of the raw. It's implicitly being moved here with the return statement here.

Reply:
^ sorry @JackCaoG it's possible that this (2nd) PR doesn't have that change, and I used to save the raw_ptr before.